src/main/java/org/apache/sysml/hops/ipa/InterProceduralAnalysis.java

```diff
@@ -543,7 +560,28 @@ else if (sb instanceof ForStatementBlock) //incl parfor
 			propagateStatisticsIntoFunctions(prog, roots, fcand, callVars, fcandSafeNNZ, unaryFcands, fnStack);
 		}
 	}
-	
+
+	/**
+	 * Propagate scalar values across DAGs.
+	 *
+	 * This replaces scalar reads and typecasts thereof with literals.
+	 *
+	 * @param roots  List of HOPs.
+	 * @param vars  Calling program's map of variables eligible for
+	 *                 propagation.
+	 * @throws HopsException
+	 */
+	private void propagateScalarsAcrossDAG(ArrayList<Hop> roots, LocalVariableMap vars)
+		throws HopsException
+	{
+		for (Hop hop : roots) {
+			try {
+				Recompiler.rReplaceLiterals(hop, vars, true);
```

Comment:
I was thinking about the need for this and I guess it leads to improvements because the size expression evaluation over dags with scalar symbol table entries (which is also applied during IPA) is limited in supported operations whereas literal replacement and constant folding is a brute force method which applies to all (including future) operations because we simply compile runtime instructions during constant folding. Bottom line: it's good to have it in IPA.